src/charm.py

```diff
@@ -64,6 +69,29 @@ def _on_config_changed(self, event) -> None:
         self.model.unit.status = MaintenanceStatus(msg)
         self.configure_workload_container(event)
 
+    def _report_visits_by_ip_action(self, event: ActionEvent) -> None:
+        """Handle the report-visits-by-ip action."""
+        results = self._report_visits_by_ip()
+        event.set_results({"ips": results})
+
+    def _report_visits_by_ip(self) -> list[(int, str)]:
+        """Report requests to nginx grouped and ordered by IP and report action result."""
+        container = self.unit.get_container(CONTAINER_NAME)
+        log_path = "/var/log/nginx/access.log"
+        with container.pull(log_path) as log_file:
+            list_to_search = log_file.readlines()
+        ip_regex = r"[0-9]+(?:\.[0-9]+){3}"
+        ip_list = []
+        current_datetime = datetime.now()
+        start_datetime = current_datetime - timedelta(minutes=20)
+        for line in list_to_search:
+            line = line.split()
+            log_datetime = datetime.strptime(line[3].lstrip("[").rstrip("]"), "%d/%b/%Y:%H:%M:%S")
+            if log_datetime >= start_datetime and re.search(ip_regex, line[0]):
+                ip_list.append(line[0])
+        results = [(len(list(group)), key) for key, group in groupby(sorted(ip_list))]
```

Comment:
This will return a list of tuples, which seems quite complex for a human readable output. Have you considered instead returning a map with `ip`: `ocurrences`?
```
{ ip: ip_list.count(ip) for ip in ip_list }
```

Reply:
I tried that, but the results aren't printed like they should: https://pastebin.canonical.com/p/Xvywypm7j8/